keccak256/src/circuit/padding.rs

```diff
@@ -0,0 +1,443 @@
+                        - byte_cur
+                        - diff_is_zero.clone().is_zero_expression
+                            * Expression::Constant(F::from(0x80)),
+                )))
+                .map(move |(name, poly)| (name, q_without_last.clone() * poly))
+        });
+        meta.create_gate("without first", |meta| {
+            let q_without_first = meta.query_selector(q_without_first);
+            let is_pad_zone_prev = meta.query_advice(is_pad_zone, Rotation::prev());
+            let is_pad_zone_cur = meta.query_advice(is_pad_zone, Rotation::cur());
+            vec![(
+                "check pad_zone",
+                q_without_first
+                    * (is_pad_zone_cur
+                        - is_pad_zone_prev
+                        - diff_is_zero.clone().is_zero_expression),
+            )]
+        });
+
+        meta.create_gate("last", |meta| {
+            let q_last = meta.query_selector(q_last);
+            let is_finalize = meta.query_advice(is_finalize, Rotation::cur());
+            let padded_byte_cur = meta.query_advice(padded_byte, Rotation::cur());
+            let byte_cur = meta.query_advice(byte, Rotation::cur());
+            vec![
+                q_last
+                    * (padded_byte_cur
+                        - byte_cur
+                        - diff_is_zero.clone().is_zero_expression
+                            * Expression::Constant(F::from(0x80))
+                        - is_finalize),
+            ]
+        });
+        Self {
+            q_all,
+            q_without_first,
+            q_without_last,
+            q_last,
+            is_finalize,
+            byte,
+            input_len,
+            acc_len,
+            diff_is_zero,
+            is_pad_zone,
+            padded_byte,
+            word_config,
+        }
+    }
+    pub fn assign_region(
+        &self,
+        layouter: &mut impl Layouter<F>,
+        is_finalize: AssignedCell<F, F>,
+        input_len_cell: AssignedCell<F, F>,
+        acc_len_cell: AssignedCell<F, F>,
+        bytes: [u8; BYTES_LEN_17_WORDS],
+    ) -> Result<[AssignedCell<F, F>; 17], Error> {
+        let diff_is_zero_chip = IsZeroChip::construct(self.diff_is_zero.clone());
+        layouter.assign_region(
+            || "padding validation",
+            |mut region| {
+                let last = BYTES_LEN_17_WORDS - 1;
+                self.q_last.enable(&mut region, last)?;
+                let mut is_pad_zone = F::zero();
+                let mut padded_bytes = [0u8; BYTES_LEN_17_WORDS];
+                for (offset, &byte) in bytes.iter().enumerate().take(BYTES_LEN_17_WORDS) {
+                    self.q_all.enable(&mut region, offset)?;
+                    if offset != 0 {
+                        self.q_without_first.enable(&mut region, offset)?;
+                    }
+                    if offset != last {
+                        self.q_without_last.enable(&mut region, offset)?;
+                    }
```

Comment:
`q_without_first` and `q_without_last` are just selectors. 
We have one padding region for each keccak_f round, no matter if they need padding or not. Each padding region has 136 rows. Some gates are enabled 1~135(without first), some are 0~134(without last), and some are 0~135(all). Gates are not conditioned on any advice values.

Reply:
>Gates are not conditioned on any advice values.

That's exactly what I meant. Advice values are not. But if you use selectors, we will need to save 3 different ProverKeys, one for each case.

My point is that the padding Config has to select first. last, or any  other case with Advice values that multiply a constant selector that is always active.


Otherwise, we need one ProverKey where the selector for `without_first ==1`, another for `without_last ==1` and so on.

Reply:
If we use an advice column to handle without_first/without_last, wouldn't we need more gates and openings for constraining the column? Besides, would 1 advice column be enough? I think we still need 2 to represent without_first and without_last?

Reply:
hmmm. Good points.

Let me bring @therealyingtong or @han0110 to the discussion. I would like to know if the `Selector` approach is ok independently of the hardcoded stuff it includes.

As for your  questions: 
1) We would need an extra gate that checks that the value is between 0-2. We can do a lookup too or whatever. Should not be too much. We can do it with Lagrange as we do in the Lookup table of permutations.
2) 1 Advice should suffice as the gates will activate/deactivate depending on the advice column value. 
3) Note that the gates will be in different regions which means that we should not need more than one.

Reply:
Not sure I follow all constraint in this PR, but using selector seems good to me since it doesn't depend on witness to be enabled, we would have static prover/verifier key, right?

Also if we use simple selectors, halo2 will optimize them into lagrange style assignment as you describe. Tho in this case these 2 won't be merge together because they seem to have overlap, but they still have chance to be merged with other simple selectors.

Reply:
Ohhh wait wait! I see. 

@ChihChengLiang correct me if I'm wrong but the selectors will activate always at the same places. No matter what the padding that needs to be done is. 

I was under the impression that this wasn';t the case and therefore, my concern.

Apologies for that. This is resolved.

Reply:
>  the selectors will activate always at the same places. No matter what the padding that needs to be done is.

@CPerezz We changed to that behavior per our out of the band discussion May 5.